src/helpers/script_helpers.sh

```diff
@@ -117,6 +145,15 @@ function run {
     preconditions "$@"
     steps > "$STEPS_FILE"
     run_steps "$STEPS_FILE" undoable
+
+    # Apply global previous-branch restoration
+    # only if we're at the git root.
+    # Any branch restoration logic for the other
+    # case needs to be handled in a custom way by
+    # the command itself
```

Comment:
I like the approach of (2), and think we should go even further with this idea. Almost every command has `echo_if_true "stash_open_changes" "$INITIAL_OPEN_CHANGES"`. And a lot of them need to preserve the Git history. These commands just provide a clean, standardized workspace to run the actual command logic safely. We should make these things dedicated, configurable services that are automatically run by the GT runtime.

```bash
# src/helpers/script_helpers.sh
function run {
  if [ "$1" = "--abort" ]; then
    ensure_abortable
    abort_command
  elif [ "$1" = "--continue" ]; then
    ensure_continuable
    ensure_no_conflicts
    continue_command
  elif [ "$1" = "--skip" ]; then
    ensure_skippable
    skip_command
  elif [ "$1" = "--undo" ]; then
    ensure_undoable
    undo_command
  else
    
    # Cleanup
    remove_step_files

    # Prepare
    preconditions "$@"
    cd_git_root                 
    save_git_history
    stash_open_changes

    # Run core command logic
    steps >> "$STEPS_FILE"
    run_steps "$STEPS_FILE" undoable

    # Restore environment
    restore_git_history
    unstash_open_changes
    cd_original_folder
  fi
}
```

Commands can configure whether they want any of the GT environment services:

```bash
# src/git-sync-fork
export stash_open_changes=false   # thanks, but I don't need changes stashed away, thank you
export preserve_git_history=true  # yes, I change Git history, please preserve it
```